### DISCUSSION TITLE
SailorHat getting started

### DISCUSSION THREAD

**Discussion by wjquigs11** [hatlabs/discussions#55] [Q&A] [answered]
Hello all, I just received my SailorHat USB Gateway. My intention is to use it to correct wind data on my rotating mast, but initially I'm just trying to get USB working. 

After uploading the code, the display shows:
SH-ESP32...
CAN: BUS-OFF
Uptime:...
RX: 0
TX: 0

However, as soon as uptime gets to 2, the display cycles and Uptime resets to 1. Does this mean the device is rebooting continuously?

Note it is connected to my Macbook but I haven't taken it to the boat yet. I can see the serial device under System Info and it shows up as a /dev/tty.usb device when it's plugged in, but SignalK does not show a connection.

**Comment by mairas** [comment, marked as answer]
Hi,

Ah, I see you figured the pinout already. The pinout is shown on the silkscreen next to the connector but you probably found it already. The wire coloring is different in the tutorial photos because the version 1 boards used JST XH connectors and I hadn't sourced pre-soldered connectors. The photos have a yellow wire instead of the standard blue one.

The device is indeed in a reboot loop, but that is more or less as intended. If you don't have 12V power connected to the N2K connector, the transceiver IC is not powered. Then, the CAN controller in the ESP32 thinks the CAN bus is faulty and enters a "BUS-OFF" state. When writing the software, I couldn't find any other way to clear the error but to reboot the chip. Hence, the reboot loop.

If you want to test the software at home, try connecting 12V and GND to the N2K header. That should keep the CAN interface alive. 

I've since seen an example of a better approach to deal with the bus off but haven't had time to implement it yet: you keep track of the CAN interface error counter and reset it before the device has time to enter BUS-OFF. An even better solution would be to rewrite the NMEA2000_esp32 library to use the ESP-IDF CAN interface functions that would allow graceful recovery from the error states.

This was probably too much information for you, but maybe the next poor soul will find it helpful. :-)

**Reply by wjquigs11** [reply]
Thanks, that makes a lot of sense (and not TMI). I thought maybe it has something to do with the N2K bus not being terminated, but no power makes sense :-)
Not only did you print the pins on the board, but it's also here (https://docs.hatlabs.fi/sh-esp32/pages/hardware) so it shouldn't be too hard to find the right ones.
Regarding power: I plan to install an I2C ADC as per randel0's guide. I assume I can use 12V/Gnd pins on the N2K connector to supply 12V to the Honeywell sensor, correct? Randel0 shows two connections to ground for the data wire from the sensor. Do those also go to the same ground or would it be better to solder them to the board somewhere else? Do I need to solder 2 wires across the Voltage Link to get power from N2K bus as per #8 on the hardware page? It does not appear that you did this in the tutorial. (Obviously I'm not a hardware engineer).
Can you point me to a good source of information on I2C? For example, it does not appear to be terminated? Can I connect multiple I2C devices to the same bus (eventually when I get mast rotation working, I would like to connect a serial interface and use one of the NMEA0183 libraries to convert GPS data from my VHF and/or AIS, thus eliminating the other ESP device on the boat).
In your code, you have several Serial.println() calls. Would that show up in the console window in Visual Studio, or would I have to explicitly connect to the dev/tty.usb device?
Thanks; I'm very excited about the possibilities of this little box.

**Reply by mairas** [reply]
Wasn't aware of that project (https://github.com/randel0/NEMA2K-Mast-Rotation-Compensator) - nice!

Let me first note that if you intend to implement a device that simultaneously acts as a USB gateway and a rotation sensor, you're going to be in a world of pain. First, the "standard" ttlappalainen/NMEA2000 library is built so that if you write N2K sentences while having the N2K gateway enabled, the sentences only end up on the N2K bus and not on the USB serial device, and that would require some additional low-level hacking to get it supported. I might be wrong but that's my understanding.

Second, the USB gateway functionality reserves the standard USB serial port, preventing you from writing debug messages easily while developing the mast rotation sensor functionality. Again, there are ways around it but things do get complex.

I really recommend using separate devices for those two functionalities.

With that out of the way, re your power output question: Instead of using the N2K pins for 12V, use the positive Voltage Link pin. That includes overvoltage and reverse polarity protection, reducing the likelihood you'll accidentally burn that expensive-looking Honeywell sensor. :-) (BTW, randel0's project description states the sensor requires 24V input voltage, but the specs state 6-24V, so 12V is indeed _much_ safer.)

I2C is _very_ widely used and there are a lot of good resources online; Wikipedia gives a good overview, and Google provides a lot of quite understandable descriptions. Usually you don't want or need to implement the low-level protocol because a lot of good Arduino libraries exist for many devices.

As for Serial port debugging, both VSCode and Arduino IDE include a built-in serial console that can be used to observe the output. Just try with some simple example code to get the hang of it.

**Reply by wjquigs11** [reply]
Hi,
I will most likely use separate devices...what you said makes sense. In the interim, I have disabled the N2K reporting on the serial port.
I tried to add an ADS1015 ADC. It initializes properly, but hangs on ads1015.readADC_SingleEnded(0)
Google revealed this possible answer: https://esp32.com/viewtopic.php?t=20799
However, adding a 5k resistor from SCL->+3.3V and from SDA->+3.3V did not change anything. I have tried using the Qwiic connection as well as going directly to the voltage output and pins 16 and 17 on the header. Do you have any recommendations?
Thanks!
William

**Reply by wjquigs11** [reply]
I connected a pot (and eventually the Honeywell sensor) to the ESP internal ADC on GPIO 36, and it seems to be working, although apparently there are some issues with the ESP32 ADC.
I have tried scanning I2C to find the ADS1015 and 1115 using this https://randomnerdtutorials.com/esp32-i2c-communication-arduino-ide/#2
But it finds no devices. I also tried bypassing the Qwiic connector and using different addresses on the ADCs.
Doesn't the voltage link provide 3.3V? Did you mean I should use the protected voltage pads for 12V to the sensor?

**Reply by mairas** [reply]
Did you change the I2C GPIO pins when running the I2C code? SH-ESP32 uses different GPIO pins from what the Arduino libraries default to.

**Reply by wjquigs11** [reply]
I started with your USB Gateway code, which has this:
#define SDA_PIN 16
#define SCL_PIN 17
Is that correct?

**Reply by mairas** [reply]
> #define SDA_PIN 16
> #define SCL_PIN 17

Yes, these are the correct ones.

**Comment by nousian** [comment]
> Let me first note that if you intend to implement a device that simultaneously acts as a USB gateway and a rotation sensor, you're going to be in a world of pain. First, the "standard" ttlappalainen/NMEA2000 library is built so that if you write N2K sentences while having the N2K gateway enabled, the sentences only end up on the N2K bus and not on the USB serial device, and that would require some additional low-level hacking to get it supported. I might be wrong but that's my understanding.

I struggled with the same issue, and solved it by ditching the usb gateway and switching it into a udp gateway. This is much nicer solution in many respects, not the least by not having to run the usb cable from the gateway. Also as you mentioned, the usb is left free for debug messages - though even for that after the very initial setup phase udp packets to debug port are way nicer in my opinion.

Incoming N2K packets are converted to N0183 $PCDIN sentences and sent as UDP packets that Signalk server digests happily. Reverse works the same way.

Relevant snippets:

```
#include <WiFi.h>
#include <WiFiUdp.h>
#include <AsyncUdp.h>
#include <N2kMessages.h>
#include <Seasmart.h>

const size_t buff_size_seasmart = 512;
char buff_seasmart[buff_size_seasmart];

void HandleStreamN2kMsg(const tN2kMsg &message) {
  n2k_message_count++;
  l
  if (WiFi.status() == WL_CONNECTED) {
    Udp.beginPacket(SIGNALK_IP, SIGNALK_PORT);

    size_t st = N2kToSeasmart(message, message.MsgTime, buff_seasmart, buff_size_seasmart);
    if (st != 0) {
      Udp.printf("%s", buff_seasmart); // works without defining length cause n0183 sentences are \0 terminated
    }

    Udp.endPacket();
  }
}
```
